tests/test_browser.go

```diff
@@ -0,0 +1,297 @@
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package tests
+
+import (
+	"context"
+	"net/http"
+	"os"
+	"strconv"
+	"testing"
+
+	"github.com/dop251/goja"
+	"github.com/grafana/xk6-browser/api"
+	"github.com/grafana/xk6-browser/chromium"
+	"github.com/oxtoacart/bpool"
+	"github.com/sirupsen/logrus"
+	"github.com/stretchr/testify/require"
+	k6common "go.k6.io/k6/js/common"
+	k6http "go.k6.io/k6/js/modules/k6/http"
+	k6lib "go.k6.io/k6/lib"
+	k6metrics "go.k6.io/k6/lib/metrics"
+	k6test "go.k6.io/k6/lib/testutils/httpmultibin"
+	k6stats "go.k6.io/k6/stats"
+	"gopkg.in/guregu/null.v3"
+)
+
+// browser is a test browser for integration testing.
+type browser struct {
+	ctx     context.Context
+	rt      *goja.Runtime
+	state   *k6lib.State
+	http    *k6test.HTTPMultiBin
+	samples chan k6stats.SampleContainer
+	api.Browser
+}
+
+// testBrowser configures and launches a chrome browser.
+// It automatically closes the browser when `t` returns.
+//
+// opts provides a way to customize the testBrowser.
+// see: withLaunchOptions for an example.
+func testBrowser(t testing.TB, opts ...interface{}) *browser {
+	ctx := context.Background()
+
+	// set default options and then customize them
+	var (
+		launchOpts         = defaultLaunchOpts()
+		enableHTTPMultiBin = false
+		enableFileServer   = false
+	)
+	for _, opt := range opts {
+		switch opt := opt.(type) {
+		case withLaunchOptions:
+			launchOpts = opt
+		case httpServerOption:
+			enableHTTPMultiBin = true
+		case fileServerOption:
+			enableFileServer = true
+			enableHTTPMultiBin = true
+		case withContext:
+			ctx = opt
+		}
+	}
+
+	// create a k6 state
+	root, err := k6lib.NewGroup("", nil)
+	require.NoError(t, err)
+
+	samples := make(chan k6stats.SampleContainer, 1000)
+
+	state := &k6lib.State{
+		Options: k6lib.Options{
+			MaxRedirects: null.IntFrom(10),
+			UserAgent:    null.StringFrom("TestUserAgent"),
+			Throw:        null.BoolFrom(true),
+			SystemTags:   &k6stats.DefaultSystemTagSet,
+			Batch:        null.IntFrom(20),
+			BatchPerHost: null.IntFrom(20),
+			// HTTPDebug:    null.StringFrom("full"),
+		},
+		Logger:         logrus.StandardLogger(),
+		Group:          root,
+		BPool:          bpool.NewBufferPool(1),
+		Samples:        samples,
+		Tags:           k6lib.NewTagMap(map[string]string{"group": root.Path}),
+		BuiltinMetrics: k6metrics.RegisterBuiltinMetrics(k6metrics.NewRegistry()),
+	}
+
+	// configure the goja VM (1)
+	rt := goja.New()
+
+	// enable the HTTP test server only when necessary
+	var testServer *k6test.HTTPMultiBin
+	if enableHTTPMultiBin {
+		testServer = k6test.NewHTTPMultiBin(t)
+		state.TLSConfig = testServer.TLSClientConfig
+		state.Transport = testServer.HTTPTransport
+
+		err = rt.Set("http", k6common.Bind(rt, new(k6http.GlobalHTTP).NewModuleInstancePerVU(), &ctx))
+		require.NoError(t, err)
+	}
+
+	// configure the goja VM (2)
```

Comment:
What do you mean?

Reply:
I mean that it's not useful to comment "goja VM (1)" and "goja VM (2)", but that it would be useful to comment _why_ we need two separate goja VMs by explaining the purpose of each one.